### ISSUE TITLE
unable to successfully restore device 

### ISSUE THREAD

**Issue by iM4CH3T3** [libimobiledevice/idevicerestore#326]
```
im4ch3t3@im4ch3t3-ubuntu:~$ idevicerestore -e '/home/im4ch3t3/Downloads/iPhone11,8,iPhone12,1_13.4.1_17E262_Restore.ipsw' 
Found device in Recovery mode
Identified device as n104ap, iPhone12,1
Extracting BuildManifest from IPSW
Product Version: 13.4.1
Product Build: 17E262 Major: 17
INFO: device serial number is C8PZ8BMRN735
Device supports Image4: true
Variant: Customer Erase Install (IPSW)
This restore will erase your device data.
################################ [ WARNING ] #################################
# You are about to perform an *ERASE* restore. ALL DATA on the target device #
# will be IRREVERSIBLY DESTROYED. If you want to update your device without  #
# erasing the user data, hit CTRL+C now and restart without -e or --erase    #
# command line switch.                                                       #
# If you want to continue with the ERASE, please type YES and press ENTER.   #
##############################################################################
> YES
Checking IPSW for required components...
All required components found in IPSW
Using cached filesystem from '/home/im4ch3t3/Downloads/iPhone11,8,iPhone12,1_13.4.1_17E262_Restore/038-03660-158.dmg'
Found ECID 3757688942002222
Getting ApNonce in recovery mode... 6a 28 71 6c 9c e5 88 68 9f fd 73 34 8d 84 5d 41 24 d5 64 9f c2 8b 4a 32 d6 55 6d de fa 53 2f 1f 
Trying to fetch new SHSH blob
Getting SepNonce in recovery mode... 8c 6e 81 5a 98 72 cc f9 6b 72 3a 4a e2 1e 34 ee fa dc 46 8f 
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received SHSH blobs
Extracting iBEC.n104.RELEASE.im4p...
Personalizing IMG4 component iBEC...
Sending iBEC (1192123 bytes)...
Recovery Mode Environment:
iBoot build-version=iBoot-5540.102.4
iBoot build-style=RELEASE
Sending RestoreLogo...
Extracting applelogo@1792~iphone.im4p...
Personalizing IMG4 component RestoreLogo...
Sending RestoreLogo (20164 bytes)...
Extracting h12_ane_fw_metis.im4p...
Personalizing IMG4 component ANE...
Sending ANE (12347740 bytes)...
Extracting aopfw-iphone12baop.im4p...
Personalizing IMG4 component AOP...
Sending AOP (1663404 bytes)...
Extracting AppleAVE2FW.im4p...
Personalizing IMG4 component AVE...
Sending AVE (1113324 bytes)...
Extracting armfw_g12p.im4p...
Personalizing IMG4 component GFX...
Sending GFX (3470444 bytes)...
Extracting adc-zelus-n104.im4p...
Personalizing IMG4 component ISP...
Sending ISP (9742284 bytes)...
Extracting t8030pmp.im4p...
Personalizing IMG4 component PMP...
Sending PMP (340396 bytes)...
Extracting 048-65173-372.dmg.trustcache...
Personalizing IMG4 component RestoreTrustCache...
Sending RestoreTrustCache (13085 bytes)...
Extracting SmartIOFirmwareT8030.im4p...
Personalizing IMG4 component SIO...
Sending SIO (1161264 bytes)...
Extracting WirelessPower.iphone12b.im4p...
Personalizing IMG4 component WCHFirmwareUpdater...
Sending WCHFirmwareUpdater (56715 bytes)...
ramdisk-size=0x20000000
Extracting 048-65173-372.dmg...
Personalizing IMG4 component RestoreRamDisk...
Sending RestoreRamDisk (104524207 bytes)...
Extracting DeviceTree.n104ap.im4p...
Personalizing IMG4 component RestoreDeviceTree...
Sending RestoreDeviceTree (222007 bytes)...
Extracting kernelcache.release.iphone12b...
Personalizing IMG4 component RestoreKernelCache...
Sending RestoreKernelCache (14940457 bytes)...
Waiting for device to enter restore mode...
About to restore device... 
Connecting now...
Connected to com.apple.mobile.restored, version 15
Device 00008030-000D59992290802E has successfully entered restore mode
Hardware Information:
BoardID: 4
ChipID: 32816
UniqueChipID: 3757688942002222
ProductionMode: true
Starting FDR listener thread
About to send RootTicket...
Sending RootTicket now...
Done sending RootTicket
Waiting for NAND (28)
Updating NAND Firmware (58)
About to send NORData...
Found firmware path Firmware/all_flash
Getting firmware manifest from build identity
Extracting LLB.n104.RELEASE.im4p...
Personalizing IMG4 component LLB...
Extracting applelogo@1792~iphone.im4p...
Personalizing IMG4 component AppleLogo...
Extracting batterycharging0@1792~iphone.im4p...
Personalizing IMG4 component BatteryCharging0...
Extracting batterycharging1@1792~iphone.im4p...
Personalizing IMG4 component BatteryCharging1...
Extracting batteryfull@2x~iphone.im4p...
Personalizing IMG4 component BatteryFull...
Extracting batterylow0@2x~iphone.im4p...
Personalizing IMG4 component BatteryLow0...
Extracting batterylow1@2x~iphone.im4p...
Personalizing IMG4 component BatteryLow1...
Extracting glyphplugin@1792~iphone-lightning.im4p...
Personalizing IMG4 component BatteryPlugin...
Extracting DeviceTree.n104ap.im4p...
Personalizing IMG4 component DeviceTree...
Extracting lowpowermode@1792~iphone-lightning.im4p...
Personalizing IMG4 component LowPowerWallet0...
Extracting recoverymode@1792~iphone-lightning.im4p...
Personalizing IMG4 component RecoveryMode...
Extracting iBoot.n104.RELEASE.im4p...
Personalizing IMG4 component iBoot...
Extracting sep-firmware.n104.RELEASE.im4p...
Personalizing IMG4 component RestoreSEP...
Extracting sep-firmware.n104.RELEASE.im4p...
Personalizing IMG4 component SEP...
Sending NORData now...
Done sending NORData
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Checking filesystems (15)
About to send FDR Trust data...
Sending FDR Trust data now...
Done sending FDR Trust Data
ERROR: Unable to receive message from FDR 0x7ff828000e20 (-2). 0/2 bytes
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Creating partition map (11)
Creating filesystem (12)
About to send filesystem...
Connected to ASR
Validating the filesystem
Filesystem validated
Sending filesystem now...
[==================================================] 100.0%
Done sending filesystem
Verifying restore (14)
[==================================================] 100.0%
Checking filesystems (15)
Creating Protected Volume (67)
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Unmounting filesystems (29)
Checking filesystems (15)
Checking filesystems (15)
Checking filesystems (15)
Checking filesystems (15)
Mounting filesystems (16)
Mounting filesystems (16)
Mounting filesystems (16)
About to send KernelCache...
Extracting kernelcache.release.iphone12b...
Personalizing IMG4 component KernelCache...
Sending KernelCache now...
Done sending KernelCache
Installing kernelcache (27)
About to send DeviceTree...
Extracting DeviceTree.n104ap.im4p...
Personalizing IMG4 component DeviceTree...
Sending DeviceTree now...
Done sending DeviceTree
Installing DeviceTree (61)
Flashing firmware (18)
[==================================================] 100.0%
Requesting FUD data (36)
Found FUD component 'ANE'
Found FUD component 'AOP'
Found FUD component 'AVE'
Found FUD component 'Ap,HapticAssets'
Found FUD component 'AudioCodecFirmware'
Found FUD component 'GFX'
Found FUD component 'ISP'
Found FUD component 'LeapHaptics'
Found FUD component 'Multitouch'
Found FUD component 'PMP'
Found FUD component 'RestoreTrustCache'
Found FUD component 'SIO'
Found FUD component 'StaticTrustCache'
Sending FUD image list
Extracting h12_ane_fw_metis.im4p...
Personalizing IMG4 component ANE...
Sending FUD data for ANE...
Extracting aopfw-iphone12baop.im4p...
Personalizing IMG4 component AOP...
Sending FUD data for AOP...
Extracting AppleAVE2FW.im4p...
Personalizing IMG4 component AVE...
Sending FUD data for AVE...
Extracting N104_HapticAssets.im4p...
Personalizing IMG4 component Ap,HapticAssets...
Sending FUD data for Ap,HapticAssets...
Extracting N104_AudioCodecFirmware.im4p...
Personalizing IMG4 component AudioCodecFirmware...
Sending FUD data for AudioCodecFirmware...
Extracting armfw_g12p.im4p...
Personalizing IMG4 component GFX...
Sending FUD data for GFX...
Extracting adc-zelus-n104.im4p...
Personalizing IMG4 component ISP...
Sending FUD data for ISP...
Extracting N104_LeapHapticsFirmware.im4p...
Personalizing IMG4 component LeapHaptics...
Sending FUD data for LeapHaptics...
Extracting N104_Multitouch.im4p...
Personalizing IMG4 component Multitouch...
Sending FUD data for Multitouch...
Extracting t8030pmp.im4p...
Personalizing IMG4 component PMP...
Sending FUD data for PMP...
Extracting 048-65173-372.dmg.trustcache...
Personalizing IMG4 component RestoreTrustCache...
Sending FUD data for RestoreTrustCache...
Extracting SmartIOFirmwareT8030.im4p...
Personalizing IMG4 component SIO...
Sending FUD data for SIO...
Extracting 038-03660-158.dmg.trustcache...
Personalizing IMG4 component StaticTrustCache...
Sending FUD data for StaticTrustCache...
Updating gas gauge software (47)
Updating gas gauge software (47)
Updating Stockholm (55)
Requesting FUD data (36)
Found FUD component 'ANE'
Found FUD component 'AOP'
Found FUD component 'AVE'
Found FUD component 'Ap,HapticAssets'
Found FUD component 'AudioCodecFirmware'
Found FUD component 'GFX'
Found FUD component 'ISP'
Found FUD component 'LeapHaptics'
Found FUD component 'Multitouch'
Found FUD component 'PMP'
Found FUD component 'RestoreTrustCache'
Found FUD component 'SIO'
Found FUD component 'StaticTrustCache'
Sending FUD image list
Updating baseband (19)
About to send BasebandData...
Sending Baseband TSS request...
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received Baseband SHSH blobs
WARNING: size mismatch when parsing MBN file. Continuing anyway.
WARNING: size mismatch when parsing MBN file. Continuing anyway.
Sending BasebandData now...
Done sending BasebandData
Updating Baseband in progress...
About to send BasebandData...
Sending Baseband TSS request...
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received Baseband SHSH blobs
WARNING: size mismatch when parsing MBN file. Continuing anyway.
WARNING: size mismatch when parsing MBN file. Continuing anyway.
Sending BasebandData now...
Done sending BasebandData
Updating Baseband completed.
Updating Rose (65)
Sending Rose TSS request...
Request URL set to https://gs.apple.com/TSS/controller?action=2
Sending TSS request attempt 1... response successfully received
Received Rose ticket
Extracting ftab.bin...
ERROR: Unable to find component node for Rap,RestoreRTKitOS
ERROR: Unable get path for 'Rap,RestoreRTKitOS' component
ERROR: restore_send_firmware_updater_data: Couldn't get Rose firmware data
ERROR: Unable to send FirmwareUpdater data
ERROR: Unable to successfully restore device
ERROR: Unable to restore device
im4ch3t3@im4ch3t3-ubuntu:~$ 
```

**Comment by nikias** [comment]
Can you try the patch of this PR: https://github.com/libimobiledevice/idevicerestore/pull/323

**Comment by nikias** [comment]
Also duplicate of #325.

**Comment by iM4CH3T3** [comment]
Will try pr323 when back home ...

**Comment by iM4CH3T3** [comment]
> Can you try the patch of this PR: https://github.com/libimobiledevice/idevicerestore/pull/323

Worked with the fix of PR #323
![Uploading 79001144...]()

**Comment by iM4CH3T3** [comment]
> > Can you try the patch of this PR: https://github.com/libimobiledevice/idevicerestore/pull/323
> 
> Worked with the fix of PR #323
> ![Uploading 79001144...]()

Do u need the whole output ?

**Comment by nikias** [comment]
Fixed it with commit f27d19e1c2e69dba271834d0c7a3a5be3a83e275.
